iiif-data-service.js

```diff
@@ -0,0 +1,346 @@
+     * @returns {Promise<Object>} Image URL, width, and height
+     */
+    async extractImageInfo(canvasData) {
+        // Support both IIIF v2 and v3 formats
+        const imgUrl = canvasData?.items?.[0]?.items?.[0]?.body?.id ?? 
+                       canvasData?.images?.[0]?.resource?.["@id"] ??
+                       canvasData?.images?.[0]?.resource?.id
+                      
+        const imgWidth = canvasData?.width
+        const imgHeight = canvasData?.height
+
+        if (!imgUrl || !imgWidth || !imgHeight) {
+            throw new Error("Missing required image data in IIIF canvas")
+        }
+
+        // Check if the image URL points to an info.json (IIIF Image API)
+        const processedImageUrl = await this.processIIIFImageUrl(imgUrl, imgWidth, imgHeight)
+
+        return { imgUrl: processedImageUrl, imgWidth, imgHeight }
+    }
+
+    /**
+     * Process IIIF Image URL and handle info.json responses
+     * @param {string} imgUrl - Original image URL
+     * @param {number} maxWidth - Maximum width from canvas
+     * @param {number} maxHeight - Maximum height from canvas
+     * @returns {Promise<string>} Processed image URL
+     */
+    async processIIIFImageUrl(imgUrl, maxWidth, maxHeight) {
+        try {
+            // First, try to fetch the URL to see if it returns info.json
+            const response = await fetch(imgUrl)
+            
+            if (!response.ok) {
+                // If fetch fails, return original URL (might be a direct image)
+                return imgUrl
+            }
+
+            const contentType = response.headers.get('content-type')
+            
+            // Check if response is JSON (indicating info.json)
+            if (contentType?.includes('application/json') || contentType?.includes('application/ld+json')) {
+                const infoData = await response.json()
+                
+                // Validate this is a IIIF Image API info.json
+                if (this.isIIIFImageInfo(infoData)) {
+                    return this.constructIIIFImageUrl(infoData, maxWidth, maxHeight)
+                }
+            }
+            
+            // If it's not JSON or not a valid info.json, return original URL
+            return imgUrl
+            
+        } catch (error) {
+            console.warn('Error processing IIIF image URL:', error)
+            // Fallback to original URL if processing fails
+            return imgUrl
+        }
+    }
+
+    /**
+     * Check if the data is a valid IIIF Image API info.json
+     * @param {Object} data - JSON data to validate
+     * @returns {boolean} True if valid IIIF Image API info
+     */
+    isIIIFImageInfo(data) {
+        // Check for IIIF Image API v2 or v3 indicators
+        return (
+            // IIIF Image API v3
+            (data.type === "ImageService3" || data["@type"] === "ImageService3") ||
+            // IIIF Image API v2
+            (data["@context"] && (
```

Comment:
this might be handled at a top level somewhere instead, but I did it a bit quick